### PR TITLE
Add prng_key kwarg to new device API

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**
The `default.qubit.jax` device can take a `prng_key`, and if it does, it returns the same results from sampling any given state across repeated measurements, rather than randomizing the sampling of the state. 

**Description of the Change:**
Allow the `seed` on `DefaultQubit2` to be a `jax.random.PRNGKey`; if it is, have the new device replicate the `DefaultQubitJax` sampling behaviour. 

**Benefits:**
You can get the same behaviour on the new device if you want a set PRNG key when using `jax`.

**Possible Drawbacks:**
Previously, the jax device would use `jax.choice` to generate samples regardless of whether or not a PRNG key was provided; the difference was whether it used a new random PRNG key each time, or the same one over and over. 

For the new DefaultQubit, if you provide a PRNG key as the seed, the behaviour will be the same as the `default.qubit.jax` device with a set `prng_key`. If you leave the seed as None or an integer, even if the interface is `jax`, it will use the `numpy.random.default_rng` to generate the samples, and not `jax.choice`.

If you want a random PRNGKey and `jax.choice`, you can fix it by reinitializing the device each time you run, with a randomly generated PRNG key as the seed.

### CODE REVIEW

**Comment by lillian542** [comment]
[sc-44393]

File: pennylane/devices/qubit/sampling.py [hunk 1]
```diff
@@ -353,3 +394,64 @@ def sample_state(
...
+            key, subkey = jax.random.split(key)
+            keys.append(subkey)
+        samples = jnp.array(
+            [
+                jax.random.choice(_key, basis_states, shape=(shots,), p=prob)
+                for _key, prob in zip(keys, probs)
+            ]
+        )
+    else:
+        samples = jax.random.choice(key, basis_states, shape=(shots,), p=probs)
```
Comment (timmysilv) [inline pennylane/devices/qubit/sampling.py:453]: this is the only section that's different (plus the key handling) - could you change the function to only be this (something like `generate_samples_jax`) and have the non-jax counterpart in a different function (like `generate_samples`) so we don't have to copy-paste the before/after stuff. I'm only asking because sometimes the before/after stuff changes, and I don't want it to be missed in the jax one (eg. I changed the type from bool8 to int64 a while back)

ok after looking at tests I realize that they might need some re-writing, so we can just do this later, but still wanna mention it.

**Review by timmysilv — APPROVED** [review]
just a small comment on docs! also, don't forget to add a changelog entry.

I had a bigger request, but we can push it to another PR since it'll make tests more awkward

File: pennylane/devices/qubit/sampling.py (hunk 1 above)
Comment (lillian542) [reply pennylane/devices/qubit/sampling.py:448]: Makes sense, I will add a shortcut story

File: pennylane/devices/default_qubit.py [hunk 2]
```diff
@@ -59,6 +59,10 @@ class DefaultQubit(Device):
             using a pool of at most ``max_workers`` processes. If ``max_workers`` is ``None``,
             only the current process executes tapes. If you experience any
             issue, say using JAX, TensorFlow, Torch, try setting ``max_workers`` to ``None``.
+        prng_key (Optional[jax.random.PRNGKey]): An optional ``jax.random.PRNGKey``. Defaults to None.
```
Comment (trbromley) [inline pennylane/devices/default_qubit.py:62]: We also have `seed`. What is the motivation for an extra argument?

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -79,6 +79,11 @@
   when at the beginning of a circuit, thus behaving like `StatePrep`.
   [(#4583)](https://github.com/PennyLaneAI/pennylane/pull/4583)
 
+* The new device API can now take a `prng_key` kwarg, to set the key for the JAX pseudo random
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:82]: Are we changing the new device API, or the new default qubit?

File: pennylane/devices/default_qubit.py (hunk 2 above)
Comment (timmysilv) [reply pennylane/devices/default_qubit.py:62]: 1. default.qubit.jax also has `prng_key`, so it's consistent there
2. Lillian first attempted to use the value for `seed` as the PRNG key input if it's an int, but it was too messy

Concern for messiness:

`seed` can be an int, but it can also be a whole bunch of other types that ultimately become a numpy RNG. this was straightforward before, because regardless of what type the user passes in, we initialized an rng and [saved that](https://github.com/PennyLaneAI/pennylane/blob/cbc8d19e6a36f0d4d38b44eabd4a8b98db677327/pennylane/devices/default_qubit.py#L151) as the device property. It's then used in a variety of places as needed. It would degrade code quality to write each of these things twice without a better abstraction, and writing that feels overkill when adding a kwarg (that was already there before) is pretty straightforward.

`prng_key` was only used in one place - generating samples. It also behaved in a strange way, which is that if you have a jax device with a PRNG key saved to it, and ran the same circuit twice, it would re-create the same samples (because it would use the same initial seed per-execution). `seed` doesn't work that way in the new device, nor should it, because this was a jax-PRNG-key-specific behaviour.

Lastly, because of interface unwrapping and other issues, it can be hard to know _when_ the user wants regular numpy RNG, or jax PRNG. This way, it's much clearer. When generating samples, do I have a PRNG key saved?
- yes? use it
- no? use numpy rng

I think that pretty much covers it 😄

File: doc/releases/changelog-dev.md (hunk 3 above)
Comment (lillian542) [reply doc/releases/changelog-dev.md:82]: good point, thanks!

**Review by mudit2812 — APPROVED** [review]

File: pennylane/devices/default_qubit.py [hunk 4]
```diff
@@ -50,11 +51,14 @@ class DefaultQubit(Device):
...
-        seed (Union[str, None, int, array_like[int], SeedSequence, BitGenerator, Generator]): A
-            seed-like parameter matching that of ``seed`` for ``numpy.random.default_rng`` or
+        seed (Union[str, None, int, array_like[int], SeedSequence, BitGenerator, Generator, jax.random.PRNGKey]): A
+            seed-like parameter matching that of ``seed`` for ``numpy.random.default_rng``, or
             a request to seed from numpy's global random number generator.
             The default, ``seed="global"`` pulls a seed from NumPy's global generator. ``seed=None``
             will pull a seed from the OS entropy.
+            If a ``jax.random.PRNGKey`` is passed as the seed, a JAX-specific sampling function using
+            ``jax.random.choice`` and the ``PRNGKey`` will be used for sampling rather than
+            ``numpy.random.default_rng``.
```
Comment (trbromley) [inline pennylane/devices/default_qubit.py:61]: Thanks, this is great!

File: pennylane/devices/default_qubit.py [hunk 5]
```diff
@@ -247,15 +263,26 @@ def execute(
...
             seeds = self._rng.integers(2**31 - 1, size=len(vanilla_circuits))
             _wrap_simulate = partial(simulate, debugger=None, interface=interface)
             with concurrent.futures.ProcessPoolExecutor(max_workers=max_workers) as executor:
-                exec_map = executor.map(_wrap_simulate, vanilla_circuits, seeds)
+                exec_map = executor.map(
+                    _wrap_simulate,
+                    vanilla_circuits,
+                    seeds,
+                    [self._prng_key] * len(vanilla_circuits),
+                )  # does this not need interface and debugger for this part?
```
Comment (albi3ro) [inline pennylane/devices/default_qubit.py:285]: interface will strictly be numpy, as backpropagation falls over with multiprocessing. Debugging is also incompatible with multiprocessing.

Comment (lillian542) [reply pennylane/devices/default_qubit.py:285]: Ah, thanks! (Also for reminding me I need to remove these comments 😅 )
